Solution HW3

kyrylo-fursov opened the pull request:
All functions are implemented. Your feedback and suggestions for improvement are welcome. Thank you for taking the time to review this pull request.

kyrylo-fursov committed:
-All tasks completed

kyrylo-fursov committed:
- Import cleaned

vitaliiStefanchakInventorsoft commented on article-analyzer/build.gradle:
```diff
@@ -0,0 +1,36 @@
+plugins {
+	id 'java'
+	id 'org.springframework.boot' version '3.2.0-M3'
```
prefer stable spring releases (like 3.1)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/ArticleAnalyzer.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public void analyze() {
+        List<Article> articles = articleRepository.fetchAllArticles(); // Fetch all articles from the default file
+        Set<String> categories = categoryService.findMostCommonCategories(articles); // Find the most common categories
+        setToJsonWriter.write(categories); // Write the categories to the default file
+        notificationService.notify(categoryService.findMostCommonCategories(articles)); // Notify subscribers about the new article
```
the code tells exactly the same as what these comments say to us (of course if the names are proper)
Remove them

![image](https://github.com/vitaliiStefanchakInventorsoft/spring-lecture/assets/103140969/100bbfee-0b93-4c9e-b975-a2a01e050850)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/model/User.java:
```diff
@@ -0,0 +1,34 @@
+import lombok.Getter;
+import org.springframework.stereotype.Component;
+
+@Component
```
why data class is marked as bean?

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/notifier/EmailNotificationService.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.context.ApplicationContext;
+import org.springframework.stereotype.Service;
+
+@Service("emailNotificationService")
```
this name is needless, spring does create the bean with the same name as is specified here

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/notifier/NotificationService.java:
```diff
@@ -0,0 +1,39 @@
+            });
+    }
+
+    // Notify a single user about the result
```
remove these comments or add java docs if it helps to understand the obscure logic inside

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,45 @@
+@Repository("articleRepository")
+public class ArticleRepository {
+    private final Gson gson;
+    private final String filePath;
```
just use @value directly here over the filed, it will be better then via constructor param

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    // Use the default file path
+    public List<Article> fetchAllArticles() {
+        return fetchAllArticles(this.filePath);
+    }
+
+    // Read all articles from the JSON file
+    public List<Article> fetchAllArticles(String filePath) {
```
Is this method used somewhere else? if not, merge with the above one

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/repository/UserRepository.java:
```diff
@@ -0,0 +1,45 @@
+@Repository("userRepository")
+public class UserRepository {
+    private final Gson gson;
+    private final String filePath;
```
same as above

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/service/CategoryService.java:
```diff
@@ -0,0 +1,60 @@
+        this.excludedWords = excludedWords;
+    }
+
+    // Return the set of categories that are most common across all articles in the default file
```
either use java docs `/* .... **/` or avoid comments at all

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/service/CategoryService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    // Split the article content into words
+    private static String[] getWords(Article article) {
```
just a rule of thumb, try to express what is written in the comment as a method name (or variable name).
If you feel that comment is needed, first think about the naming

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/service/SetToJsonWriter.java:
```diff
@@ -0,0 +1,31 @@
+import java.io.IOException;
+import java.util.Set;
+
+@Service("setToJsonWriter")
```
remove all of these custom bean names across other classes, spring does create bean with exactly the name is set here

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/spring/articleanalyzer/service/CategoryService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    // Return the set of categories that are most common across all articles
+    public Set<String> findMostCommonCategories(List<Article> articles) {
+        return articles.stream()
+                .map(this::findMostCommonCategories)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+    }
```
![LikeLikeItGIF](https://github.com/vitaliiStefanchakInventorsoft/spring-lecture/assets/103140969/934f6c6b-62c1-4c2d-b6db-f12399070510)

vitaliiStefanchakInventorsoft reviewed (commented):
nice work, just few minor comments, and please commit categories file as well

kyrylo-fursov committed:
- Spring Boot version changed to 3.1.4
- unnecessary comments removed
- custom bean names removed
- categories.json committed
- Value injection utilizes field injection instead of constructor injection

kyrylo-fursov commented:
Thank you for the review! Already committed changes.
Now I will be more selective in my comments)
Why it is recommended to use @Value annotation directly above the field and not in the constructor? As far as I understand, it is recommended to favor constructor injection, and use setter only when our field is not required.

vitaliiStefanchakInventorsoft commented:
> Why it is recommended to use @value annotation directly above the field and not in the constructor? As far as I understand, it is recommended to favor constructor injection, and use setter only when our field is not required.

yep, makes sense

vitaliiStefanchakInventorsoft commented:
![ConsiderItDoneItsDoneGIF](https://github.com/vitaliiStefanchakInventorsoft/spring-lecture/assets/103140969/336bc199-b337-479c-b9d3-69991df65f8b)
